ddht/v5_1/alexandria/partials/proof.py

```diff
@@ -344,6 +345,45 @@ def get_proven_data_segments(self, length: int) -> Iterable[Tuple[int, bytes]]:
         if not length:
             yield 0, b""
 
+    def merge(self, other: "Proof") -> "Proof":
+        if self.hash_tree_root != other.hash_tree_root:
+            raise ValueError(
+                f"Mismatched roots: {self.hash_tree_root.hex()} !+ {other.hash_tree_root.hex()}"
+            )
+
+        all_self_elements_by_path = {el.path: el for el in self.elements}
+        all_other_elements_by_path = {el.path: el for el in other.elements}
+
+        all_self_paths = {el.path for el in self.elements}
+        all_other_paths = {el.path for el in other.elements}
+
+        common_paths = all_self_paths & all_other_paths
+        paths_to_merge = sorted(all_self_paths ^ all_other_paths)
+
+        merged_paths = merge_paths(*paths_to_merge)
+
+        common_elements = tuple(
+            all_self_elements_by_path[path] for path in common_paths
+        )
+        merged_elements_from_self = tuple(
+            all_self_elements_by_path[path]
+            for path in merged_paths
+            if path in all_self_elements_by_path
+        )
+        merged_elements_from_other = tuple(
+            all_other_elements_by_path[path]
+            for path in merged_paths
+            if path in all_other_elements_by_path
+        )
+
+        all_merged_elements = (
+            common_elements + merged_elements_from_self + merged_elements_from_other
+        )
+
+        proof = Proof(self.hash_tree_root, all_merged_elements, self.sedes)
+        validate_proof(proof)
+        return proof
```

Comment:
Context gathering: So this merge applies to two partial proofs. What is the upper bounds of the number of partial proofs required for some data request? Is there some workflow for more than two proofs or is that a recursive use of this functionality, or something else?

Reply:
Data requests in alexandria end up transferring about 750 bytes of data (and about 400 bytes of overhead from the merkle stuff).  According to Etherscan and some very quick sampling, ethereum blocks are about 45kb which means it'll take about 60 messages to transfer all the data, which means 60 partial proofs that end up getting re-assembled.  You can see the preliminary code that does this here: 

https://github.com/ethereum/ddht/pull/165/files#diff-e8d0d620a600d82775b4794f8378624f90f1a7c0e9ed3e692da2a39e2a10de89R286-R480

Reply:
You can use this to extrapolate some of the other data payloads.  Headers will always be sent in a single message because they are not big enough to need to split them up.  Individual transactions will typically fit in a single message (and thus won't require a proof) but sometimes will need to be split up.  etc, etc.